### PR TITLE
Implement public client application global cache

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Fixes an issue where driver continues to prompt for credentials when requesting access token for AAD authentication, mainly visible when using Interactive authentication.

### CODE REVIEW

**Comment by karinazhou** [comment]
Though the testing for this requires human interaction, can we have some comments or sample code somewhere about how to validate it manually?

**Comment by DavoudEshtehari** [comment]
Assume that there are two active tokens with different access to two databases; The first token with index 0 in the accounts has access to `database_1` and the second token has access to `database_2`.
In the next connection with the initial catalog = database_2 always we return the first token if we don't specify the user.
Maybe we need more data to pick the correct token.

**Comment by David-Engel** [comment]
> Assume that there are two active tokens with different access to two databases; The first token with index 0 in the accounts has access to `database_1` and the second token has access to `database_2`.
> In the next connection with the initial catalog = database_2 always we return the first token if we don't specify the user.
> Maybe we need more data to pick the correct token.

The scope of the token is database.windows.net, not the specific database. So using the same token for two databases will be fine. The token identifies the user to the database and database ACL settings determine whether that user is authorized to access the database. There really should never be two tokens in the situation you described. The first token would have been used when the second was requested.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs [hunk 1]
```diff
@@ -320,5 +297,168 @@ private class CustomWebUi : ICustomWebUi
...
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj == null)
+                {
+                    return false;
+                }
+
+                PublicClientAppKey pcaKey = obj as PublicClientAppKey;
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs:400]: If this conversion fails, a null ref exception will be thrown below. This check should be sufficient.

```
if (!(obj is PublicClientAppKey pcaKey))
{
    return false;
}
```

**Review by karinazhou — APPROVED** [review]

**Review by JRahnama — APPROVED** [review]

**Comment by DavoudEshtehari** [comment]
For the first named-pipe connection the exception is:

**Review by DavoudEshtehari — APPROVED** [review]

**Pull request merged**